planner/core/find_best_task.go

```diff
@@ -450,18 +451,31 @@ func compareBool(l, r bool) int {
 	return 1
 }
 
+func compareIndexBack(lhs, rhs *candidatePath) (int, bool) {
+	result := compareBool(lhs.isSingleScan, rhs.isSingleScan)
+	if result == 0 && !lhs.isSingleScan {
+		// if both lhs and rhs need to access table after IndexScan, we use the set of columns that occurred in IndexFilters
+		// to compare how many table rows will be accessed.
+		return compareColumnSet(lhs.indexFiltersColSet, rhs.indexFiltersColSet)
+	}
+	return result, true
+}
```

Comment:
Is there any test case that can cover this rule?

Reply:
Yes. https://github.com/xuyifangreeneyes/tidb/blob/116e60b8ad38f7ebfa5ea68b09e87054723a9394/planner/core/logical_plan_test.go#L1703-L1710 In the two cases `f_g` is better than `f` due to this rule.